File: Archiver/modResizeForm.vb
Imports System.Windows.Forms

'Now for how to use it, in the Form_Load Event you add a call to:

'    * GetLocation
'    * ResizeForm

'And set the FontSize property of your controls, like this:
'Private Sub Form_Load()
'    GetLocation(Me)
'    CenterForm(Me)
'    ResizeForm(Me)

'    lblInstructions.Font = SetFontSize()
'End Sub

'This Code Module also offers another feature. If someone 
'were the resize your form while having the application running, 
'if you put a call to ResizeControls in the Form_Resize Event, '
'this will take care of resizing everthing:

'CODE

'Private Sub Form_Resize()
'    ResizeControls(Me)
'    lblInstructions.FontSize = SetFontSize()
'End Sub


Module modResizeForm
    Private Structure ControlStruct
        Dim Index As Integer
        Dim Name As String
        Dim Left As Integer
        Dim Top As Integer
        Dim width As Integer
        Dim height As Integer
    End Structure

    Private List() As ControlStruct
    Private curr_obj As Object
    Private iHeight As Integer
    Private iWidth As Integer
    Private x_size As Double
    Private y_size As Double

    Dim sControls As ControlStruct()

    Public Sub ResizeForm(ByVal frm As Form, ByVal isMdiChild As Boolean)
        If isMdiChild = True Then
            'Set the forms height
            frm.Height = frm.Height / 2
            'Set the forms width
            frm.Width = frm.Width / 2
            'Resize all of the controls
            'based on the forms new size
        Else
            frm.Width = Screen.PrimaryScreen.Bounds.Width / 2
            frm.Height = Screen.PrimaryScreen.Bounds.Height / 2
        End If
        ''Set the forms height
        'frm.Height = Screen.height / 2
        ''Set the forms width
        'frm.Width = Screen.width / 2
        ''Resize all of the controls
        ''based on the forms new size
        ResizeControls(frm)
    End Sub
    Function CountOpenForms() As Integer
        Dim I As Integer = 0
        Dim frm As Form

        For Each frm In My.Application.OpenForms
            Console.WriteLine(frm.Name.ToString)
            I += 1
        Next
        Return I
    End Function
    Public Sub ResizeControls(ByRef frm As Form)
        '** WDM Put this back when perfected
        'Dim DoNotUse As Boolean = True
        'If DoNotUse = True Then
        '    Return
        'End If

        Dim NoResize As Boolean = True
        If NoResize = True Then
            Return
        End If

        Dim iForms As Integer = CountOpenForms()

        If iForms > 3 Then
            Return
        End If


        Dim i As Integer
        '   Get ratio of initial form size to current form size
        x_size = frm.Height / iHeight
        y_size = frm.Width / iWidth

        'Loop though all the objects on the form
        'Based on the upper bound of the # of controls
        For i = 0 To UBound(List)
            'Grad each control individually
            For Each curr_obj In frm.Controls
                'Check to make sure its the right control
                If curr_obj.TabIndex = List(i).Index Then
                    'Then resize the control
                    With curr_obj
                        .Left = List(i).Left * y_size
                        .width = List(i).width * y_size
                        .height = List(i).height * x_size
                        .Top = List(i).Top * x_size
                        'Console.WriteLine(.name + " @ " + .GetType.FullName.ToString)
                        Console.WriteLine(.name + " @ " + .GetType.Name.ToString)
Dim xType AS String  = .GetType.ToString 
                        If .GetType.Name.ToString.Equals("GroupBox") Then
                            'ResizeGroupBox(curr_obj)
                        Else
                            Try
                                'Me.RatesDataGridView.Font.Size=8
                                'me.ratesdatagridview.font = new Font("Arial", 10, FontStyle.Bold) (the enum might be wrong, can't remember).
                                Dim NewFontSize As Integer = SetFontSize()
                                If NewFontSize = 0 Then
                                    NewFontSize = 8.5
                                End If
                                .font = New Font("Arial", NewFontSize, FontStyle.Regular)
                                Console.WriteLine("Set font for " + .name + " : ".GetType.ToString)
                            Catch ex As Exception
                                '** Nobody cares if the font is not resized
                                Console.WriteLine("Failed to set font for " + .name + " : ".GetType.ToString)
                            End Try
                        End If
                    End With
                End If
                'Get the next control
            Next curr_obj
        Next i
    End Sub
    Public Sub ResizeGroupBox(ByVal GRP As GroupBox)
        '** WDM Put this back when perfected
        'Dim DoNotUse As Boolean = True
        'If DoNotUse = True Then
        '    Return
        'End If

        Dim iForms As Integer = CountOpenForms()

        If iForms > 3 Then
            Return
        End If


        Dim i As Integer
        '   Get ratio of initial form size to current form size
        x_size = GRP.Height / iHeight
        y_size = GRP.Width / iWidth

        'Loop though all the objects on the form
        'Based on the upper bound of the # of controls
        For i = 0 To UBound(List)
            'Grad each control individually
            For Each curr_obj In GRP.Controls
                'Check to make sure its the right control
                If curr_obj.TabIndex = List(i).Index Then
                    'Then resize the control
                    With curr_obj
                        .Left = List(i).Left * y_size
                        .width = List(i).width * y_size
                        .height = List(i).height * x_size
                        .Top = List(i).Top * x_size
                        Console.WriteLine(.name + " : ".GetType.ToString)
                        Try
                            'Me.RatesDataGridView.Font.Size=8
                            'me.ratesdatagridview.font = new Font("Arial", 10, FontStyle.Bold) (the enum might be wrong, can't remember).
                            Dim NewFontSize As Integer = SetFontSize()
                            If NewFontSize = 0 Then
                                NewFontSize = 8.5
                            End If
                            .font = New Font("Arial", NewFontSize, FontStyle.Regular)
                            Console.WriteLine("Set font for " + .name + " : ".GetType.ToString)
                        Catch ex As Exception
                            '** Nobody cares if the font is not resized
                            Console.WriteLine("Failed to set font for " + .name + " : ".GetType.ToString)
                        End Try

                    End With
                End If
                'Get the next control
            Next curr_obj
        Next i
    End Sub
    Public Sub GetLocation(ByVal frm As Form)
        Dim i As Integer
        ReDim List(0)
        '   Load the current positions of each object into a user defined type array.
        '   This information will be used to rescale them in the Resize function.

        'Loop through each control
        Try
            For Each curr_obj In frm.Controls
                'Resize the Array by 1, and preserve
                'the original objects in the array
                ReDim Preserve List(i)
                With List(i)
                    .Name = curr_obj.name
                    .Index = curr_obj.TabIndex
                    .Left = curr_obj.Left
                    .Top = curr_obj.Top
                    .width = curr_obj.Width
                    .height = curr_obj.Height
                    'Console.WriteLine(frm.Name + " : " + .Name + " : " + .GetType.ToString)
                    'If Mid(.Name, 1, 2).Equals("dg") Then
                    '    Console.WriteLine("DataGridView")
                    'End If
                End With
                i = i + 1
            Next curr_obj

        Catch ex As Exception
            Console.WriteLine(ex.Message)
        End Try
        
        '   This is what the object sizes will be compared to on rescaling.
        iHeight = frm.Height
        iWidth = frm.Width
    End Sub


    Public Function SetFontSize() As Integer
        'Make sure x_size is greater than 0
        If Int(x_size) > 0 Then
            'Set the font size
            SetFontSize = Int(x_size * 8)
        Else
            Return 10
        End If
    End Function


End Module

File: Archiver/modResizeForm.cs
using System;
using System.Drawing;
using global::System.Windows.Forms;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;
using MODI;

namespace EcmArchiver
{

    // Now for how to use it, in the Form_Load Event you add a call to:

    // * GetLocation
    // * ResizeForm

    // And set the FontSize property of your controls, like this:
    // Private Sub Form_Load()
    // GetLocation(Me)
    // CenterForm(Me)
    // ResizeForm(Me)

    // lblInstructions.Font = SetFontSize()
    // End Sub

    // This Code Module also offers another feature. If someone 
    // were the resize your form while having the application running, 
    // if you put a call to ResizeControls in the Form_Resize Event, '
    // this will take care of resizing everthing:

    // CODE

    // Private Sub Form_Resize()
    // ResizeControls(Me)
    // lblInstructions.FontSize = SetFontSize()
    // End Sub


    static class modResizeForm
    {
        private struct ControlStruct
        {
            public int Index;
            public string Name;
            public int Left;
            public int Top;
            public int width;
            public int height;
        }

        private static ControlStruct[] List;
        private static object curr_obj;
        private static int iHeight;
        private static int iWidth;
        private static double x_size;
        private static double y_size;
        private static ControlStruct[] sControls;

        public static void ResizeForm(Form frm, bool isMdiChild)
        {
            if (isMdiChild == true)
            {
                // Set the forms height
                frm.Height = (int)(frm.Height / 2d);
                // Set the forms width
                frm.Width = (int)(frm.Width / 2d);
            }
            // Resize all of the controls
            // based on the forms new size
            else
            {
                frm.Width = (int)(Screen.PrimaryScreen.Bounds.Width / 2d);
                frm.Height = (int)(Screen.PrimaryScreen.Bounds.Height / 2d);
            }
            // 'Set the forms height
            // frm.Height = Screen.height / 2
            // 'Set the forms width
            // frm.Width = Screen.width / 2
            // 'Resize all of the controls
            // 'based on the forms new size
            ResizeControls(ref frm);
        }

        public static int CountOpenForms()
        {
            int I = 0;
            foreach (Form frm in My.MyProject.Application.OpenForms)
            {
                Console.WriteLine(frm.Name.ToString());
                I += 1;
            }

            return I;
        }

        public static void ResizeControls(ref Form frm)
        {
            // ** WDM Put this back when perfected
            // Dim DoNotUse As Boolean = True
            // If DoNotUse = True Then
            // Return
            // End If

            bool NoResize = true;
            if (NoResize == true)
            {
                return;
            }

            int iForms = CountOpenForms();
            if (iForms > 3)
            {
                return;
            }

            int i;
            // Get ratio of initial form size to current form size
            x_size = frm.Height / (double)iHeight;
            y_size = frm.Width / (double)iWidth;

            // Loop though all the objects on the form
            // Based on the upper bound of the # of controls
            var loopTo = Information.UBound(List);
            for (i = 0; i <= loopTo; i++)
            {
                // Grad each control individually
                foreach (var currentCurr_obj in frm.Controls)
                {
                    curr_obj = currentCurr_obj;
                    // Check to make sure its the right control
                    if (Conversions.ToBoolean(Operators.ConditionalCompareObjectEqual(curr_obj.TabIndex, List[i].Index, false)))
                    {
                        // Then resize the control
                        {
                            var withBlock = curr_obj;
                            withBlock.Left = List[i].Left * y_size;
                            withBlock.width = List[i].width * y_size;
                            withBlock.height = List[i].height * x_size;
                            withBlock.Top = List[i].Top * x_size;
                            // Console.WriteLine(.name + " @ " + .GetType.FullName.ToString)
                            Console.WriteLine(Operators.AddObject(Operators.AddObject(withBlock.name, " @ "), withBlock.GetType().Name.ToString()));
                            string xType = withBlock.GetType().ToString();
                            if (withBlock.GetType().Name.ToString().Equals("GroupBox"))
                            {
                            }
                            // ResizeGroupBox(curr_obj)
                            else
                            {
                                try
                                {
                                    // Me.RatesDataGridView.Font.Size=8
                                    // me.ratesdatagridview.font = new Font("Arial", 10, FontStyle.Bold) (the enum might be wrong, can't remember).
                                    int NewFontSize = SetFontSize();
                                    if (NewFontSize == 0)
                                    {
                                        NewFontSize = 8.5;
                                    }

                                    withBlock.font = new Font("Arial", NewFontSize, FontStyle.Regular);
                                    Console.WriteLine(Operators.AddObject(Operators.AddObject("Set font for ", withBlock.name), " : ".GetType().ToString()));
                                }
                                catch (Exception ex)
                                {
                                    // ** Nobody cares if the font is not resized
                                    Console.WriteLine(Operators.AddObject(Operators.AddObject("Failed to set font for ", withBlock.name), " : ".GetType().ToString()));
                                }
                            }
                        }
                    }
                    // Get the next control
                }
            }
        }

        public static void ResizeGroupBox(GroupBox GRP)
        {
            // ** WDM Put this back when perfected
            // Dim DoNotUse As Boolean = True
            // If DoNotUse = True Then
            // Return
            // End If

            int iForms = CountOpenForms();
            if (iForms > 3)
            {
                return;
            }

            int i;
            // Get ratio of initial form size to current form size
            x_size = GRP.Height / (double)iHeight;
            y_size = GRP.Width / (double)iWidth;

            // Loop though all the objects on the form
            // Based on the upper bound of the # of controls
            var loopTo = Information.UBound(List);
            for (i = 0; i <= loopTo; i++)
            {
                // Grad each control individually
                foreach (var currentCurr_obj in GRP.Controls)
                {
                    curr_obj = currentCurr_obj;
                    // Check to make sure its the right control
                    if (Conversions.ToBoolean(Operators.ConditionalCompareObjectEqual(curr_obj.TabIndex, List[i].Index, false)))
                    {
                        // Then resize the control
                        {
                            var withBlock = curr_obj;
                            withBlock.Left = List[i].Left * y_size;
                            withBlock.width = List[i].width * y_size;
                            withBlock.height = List[i].height * x_size;
                            withBlock.Top = List[i].Top * x_size;
                            Console.WriteLine(Operators.AddObject(withBlock.name, " : ".GetType().ToString()));
                            try
                            {
                                // Me.RatesDataGridView.Font.Size=8
                                // me.ratesdatagridview.font = new Font("Arial", 10, FontStyle.Bold) (the enum might be wrong, can't remember).
                                int NewFontSize = SetFontSize();
                                if (NewFontSize == 0)
                                {
                                    NewFontSize = 8.5;
                                }

                                withBlock.font = new Font("Arial", NewFontSize, FontStyle.Regular);
                                Console.WriteLine(Operators.AddObject(Operators.AddObject("Set font for ", withBlock.name), " : ".GetType().ToString()));
                            }
                            catch (Exception ex)
                            {
                                // ** Nobody cares if the font is not resized
                                Console.WriteLine(Operators.AddObject(Operators.AddObject("Failed to set font for ", withBlock.name), " : ".GetType().ToString()));
                            }
                        }
                    }
                    // Get the next control
                }
            }
        }

        public static void GetLocation(Form frm)
        {
            var i = default(int);
            List = new ControlStruct[1];
            // Load the current positions of each object into a user defined type array.
            // This information will be used to rescale them in the Resize function.

            // Loop through each control
            try
            {
                foreach (var currentCurr_obj in frm.Controls)
                {
                    curr_obj = currentCurr_obj;
                    // Resize the Array by 1, and preserve
                    // the original objects in the array
                    Array.Resize(ref List, i + 1);
                    {
                        var withBlock = List[i];
                        withBlock.Name = Conversions.ToString(curr_obj.name);
                        withBlock.Index = Conversions.ToInteger(curr_obj.TabIndex);
                        withBlock.Left = Conversions.ToInteger(curr_obj.Left);
                        withBlock.Top = Conversions.ToInteger(curr_obj.Top);
                        withBlock.width = Conversions.ToInteger(curr_obj.Width);
                        withBlock.height = Conversions.ToInteger(curr_obj.Height);
                        // Console.WriteLine(frm.Name + " : " + .Name + " : " + .GetType.ToString)
                        // If Mid(.Name, 1, 2).Equals("dg") Then
                        // Console.WriteLine("DataGridView")
                        // End If
                    }

                    i = i + 1;
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
            }

            // This is what the object sizes will be compared to on rescaling.
            iHeight = frm.Height;
            iWidth = frm.Width;
        }

        public static int SetFontSize()
        {
            int SetFontSizeRet = default;
            // Make sure x_size is greater than 0
            if (Conversion.Int(x_size) > 0d)
            {
                // Set the font size
                SetFontSizeRet = (int)Conversion.Int(x_size * 8d);
            }
            else
            {
                return 10;
            }

            return SetFontSizeRet;
        }
    }
}